File: solutions/_547_number_of_provinces/solution.php

<?php

class Solution {

    /**
     * @param Integer[][] $isConnected
     * @return Integer
     */
    function findCircleNum($isConnected) {
        $numCities = count($isConnected);
        $visited = [];
        $numProvinces = 0;

        for ($i = 0; $i < $numCities; $i++) {
            if (!isset($visited[$i])) {
                $this->dfs($isConnected, $visited, $i);
                $numProvinces++;
            }
        }

        return $numProvinces;
    }

    private function dfs($isConnected, &$visited, $i) {
        $visited[$i] = true;
        for ($j = 0; $j < count($isConnected); $j++) {
            if ($isConnected[$i][$j] === 1 && !isset($visited[$j])) {
                $this->dfs($isConnected, $visited, $j);
            }
        }
    }
}

/*
https://leetcode.com/problems/number-of-provinces/
*/

File: solutions/_547_number_of_provinces/solution.py
from typing import List

class Solution:
    def findCircleNum(self, isConnected: List[List[int]]) -> int:
        n = len(isConnected)
        visited = [False] * n
        count = 0

        for i in range(n):
            if not visited[i]:
                self.dfs(isConnected, visited, i)
                count += 1

        return count

    def dfs(self, isConnected: List[List[int]], visited: List[bool], i: int) -> None:
        visited[i] = True
        for j in range(len(isConnected)):
            if isConnected[i][j] == 1 and not visited[j]:
                self.dfs(isConnected, visited, j)


# https://leetcode.com/problems/number-of-provinces/
